README.md
## Mac OS / Debian / Ubuntu

##### INTRA HOUSE

`curl -sL https://git.io/JYNjw | sudo -E bash -s ru`  

##### INTRA SCADA

`curl -sL https://git.io/JYAeq | sudo -E bash -s ru`


## Windows

##### INTRA HOUSE

cmd: `powershell -NoProfile -InputFormat None -ExecutionPolicy Bypass -Command "$l='ru';iex ((New-Object System.Net.WebClient).DownloadString('https://git.io/JOe18'))"`  

##### INTRA SCADA

cmd: `powershell -NoProfile -InputFormat None -ExecutionPolicy Bypass -Command "$l='ru';iex ((New-Object System.Net.WebClient).DownloadString('https://git.io/JYpbW'))"`  



CHANGELOG.md

# development
- Добавлена функционал промежуточных beta версий. 

# 5.5.36 15-07-2021
- В редакторе визуальных элементов при показе дерева устройств, если нет привязки, позиционирование выполняется на последнее устройство, для которого была выполнена привязка.
- Для операции gotoLayout изменены имена свойств (container_id, device_id). Привязки в старом формате работать не будут
- Исправлен запрос trend для устройства-контекста

# 5.5.35 14-07-2021
- Добавлена функция "Перепривязать" в редакторе визуальных элементов. Выполняет привязку к устройству по аналогии с уже сделанными привязками. 
  Механизм использования:
  - выполнить привязки свойств устройства (и команд) к элементу визуализации (шаблону, примитиву,..)   
  - копировать элемент
  - в popup меню по правой кнопке мыши выбрать операцию "Перепривязать", в диалоге выбрать устройство
  Будут выполнены привязки к свойствам выбранного устройства. Операция также может быть выполнена для группы элементов
  
- Реализован механизм отправки значений свойств устройств на плагины по подписке (публикация данных системы, например, по mqtt) 
- Выпущена версия плагина mqtt 5.5.5
  - Исправлена ошибка при отправке команд управления
  - Восстановлена вкладка "Расширения" для формирования списка устройств для публикации с указанием qos, retain 
  - Публикация данных из списка "Расширения" при получении данных от сервера по подписке
  - Изменения при формировании каналов, добавленных процедурой сканирования  

# 5.5.34 12-07-2021
- Реализован механизм переключения между экранами с изменением содержимого контейнеров и устройств-контекстов. 
- Технические изменения в PM 

# 5.5.33 09-07-2021
- Добавлен протокол обмена и диалог для сканирования каналов (для плагинов, имеющих такую функцию).
  На текущий момент сканирование реализовано в плагине mqtt
- Добавлен раздел Скрипты визуализации и механизм запуска скриптов. 
  Эти скрипты используют контекст текущего экрана (например, для переключения экранов/фреймов в зависимости от локальных переменных или элементов экрана)
- Добавлена возможности по аналогии с диалогом в привязках контейнера использовать опцию "Любое устройство". 
  Привязка к фактическому устройству выполняется при установке на экран, или при выполнении команды переключения экрана/фрейма, или с помощью скрипта визуализации  

# 5.5.31 23-06-2021
- Добавлен новый виджет Chart Bar 
- Изменения в механизме отладчика плагина
- Добавлена выгрузка лога плагина
- Добавлен import/export devicedb через csv
- Добавлена обработка манифеста для dbagent (ext свойства)
- Исправлена ошибка: при копировании устройства, содержащего правила записи в БД, запись в БД начиналась только после редактирования правила или после перезагрузки
- Исправлена ошибка: при изменении типа устройства оставались свойства предыдущего типа в правилах записи в БД и в списке свойств для визуализации
- Исправлена ошибка: некорректный набор изображений при выборе изображения в диалоге при работе через p2p

# 5.5.30 23-06-2021
- Добавлен фильтр в отладчике

# 5.5.28 22-06-2021
- Доработана форма Плагины на дашборде
- Отдельный лог для каждого плагина. Настройка уровня сообщений для логирования

# 5.5.27 21-06-2021
- Добавлена вкладка Плагины на дашборд для обновления и установки плагинов
- Изменены структуры для manifest.json: плагин определяет popup для каналов 

# 5.5.26 18-06-2021
- Добавлена возможность копирования экземпляра плагина включая каналы
- Изменен манифест плагинов: manifest.json
- Добавлена возможность в манифесте задавать структуру каналов (папка, узел, канал, канал для чтения, канал для записи)  
- Новая версия системного плагина sqlite (модуль calculate - расчет времени работы)

# 5.5.24 15-06-2021
- Изменения связанные с масштабированием графиков 

# 5.5.23 11-06-2021
- Выпуск десктопной версии клиента 

# 5.5.22 09-06-2021
- Новая версия системного плагина p2p 

# 5.5.21 09-06-2021
- Добавлен multichart для графиков в IntraHouse при наличии лицензии

# 5.5.20 09-06-2021
- Обновление версии в системных индикаторах (процессах) после обновления плагина
- Остановленный плагин теперь не стартует при перезапуске сервера 
- Исправлена ошибка, связанная с calc обработчиками
- Исправлена закачка плагинов для Windows

# 5.5.19 07-06-2021
- Режим Эксперт теперь устанавливается не для проекта, а для учетной записи
- Операции, доступные только для пользователей-экспертов: 
   - webconsole (дашборд);
   - Подключение внешних модулей (настройка текущего проекта)
- Восстановлен пункт меню "БД" в админке IntraHouse 
- Пункт меню "Аналитика" в админке IntraHouse - опционально
- При загрузке плагина проверяется версия. Устанавливаются плагины только 5 версии
- В bug-report добавлены данные о версии системы и проекта
- Добавлена кнопка Сохранить при редактировании/перезагрузке проектов
- В новый проект добавляются типы устройств, шаблоны, диалоги, изображения из демо проекта
  Чтобы эта опция работала, нужно установить систему через curl 
- Отдельное окно отладчика для плагина

# 5.5.18 01-06-2021
- Увеличен таймаут (ping-pong) для потока, запускающего сниппеты с 1 до 10 сек (Wirenboard не успевает на старте) 

# 5.5.16 01-06-2021
- Добавлен системный плагин webconsole. Теперь консоль встроена в админку
- Доработаны обработчики типа - циклически и по расписанию
- Добавлена функция устройства для присваивания значений нескольким свойствам одновременно: device.setValues({prop1:val1, prop2:val2,...})
- Исправлена ошибка загрузки/выгрузки пакетов, плагинов, проектов под Windows
- Исправлена ошибка перезагрузки сервера с первой страницы дашборда, внесенная в v5.5.15

# beta 5.5.15 28-05-2021
- Добавлена отдельная вкладка для обновления системы с авторизацией
- На первой странице дашборда кнопка обновления убрана

# beta 5.5.2 12-05-2021

test.sh

#!/bin/bash


echo -e "\033[0;31m"

#-------------- functions
function config {
   pwd=$(pwd)
   name_service="ih-v5"
   root="/opt/$name_service"
   temp="$root/temp"
   node="$root/node"
   url="https://github.com/intrahouseio/ih-v5/releases/download/v0.0.0"
}

function clear {
  rm -fr $root
  mkdir -p $root
  mkdir -p $temp
  mkdir -p $node
  cd $root
  
}

function logo {
echo -e "\033[0;34m"
cat <<\EOF
   ██╗██╗  ██╗      ███████╗██╗   ██╗███████╗████████╗███████╗███╗   ███╗███████╗
   ██║██║  ██║      ██╔════╝╚██╗ ██╔╝██╔════╝╚══██╔══╝██╔════╝████╗ ████║██╔════╝
   ██║███████║█████╗███████╗ ╚████╔╝ ███████╗   ██║   █████╗  ██╔████╔██║███████╗
   ██║██╔══██║╚════╝╚════██║  ╚██╔╝  ╚════██║   ██║   ██╔══╝  ██║╚██╔╝██║╚════██║
   ██║██║  ██║      ███████║   ██║   ███████║   ██║   ███████╗██║ ╚═╝ ██║███████║
   ╚═╝╚═╝  ╚═╝      ╚══════╝   ╚═╝   ╚══════╝   ╚═╝   ╚══════╝╚═╝     ╚═╝╚══════╝
                                                                                                                                    
                         Software for Automation Systems                          
-----------------------------------------------------------------------------------
EOF
echo -e "\033[0;31m"
}

function install {
 # echo -e "\033[0;33m"
 # echo -e "INSTALL:\033[0;34m"
  echo -e "\033[0m"
  sudo sudo "$root/node/bin/node" "$root/script.js"

}

function files {
  echo -e "\033[0;33m"
  echo -e "Get loader:\033[0;34m"

  if [ ! -f "$pwd/script.js" ]; then
      curl -Lo node.tar.gz "$url/node-$platform-$arch.tar.gz"  #script not found
      curl -Lo script.js "https://raw.githubusercontent.com/intrahouseio/ih-v5/main/script.js"
    else
      curl -Lo node.tar.gz "$url/node-$platform-$arch.tar.gz" #script found
      cp "$pwd/script.js" "$root/script.js"
  fi
  
  echo -e "\033[0;33m"
  echo -e "Extract loader:\033[0;34m"
  echo ""

  while true;do echo -n .;sleep 1;done &
  cd ./node
  tar xf ./../node.tar.gz --strip 1 
  cd ./../
  kill $!; trap 'kill $!' SIGTERM
  echo "ok"
}

function run {
  config
  # clear
  logo
  # files
  cp "$pwd/script.js" "$root/script.js"
  install
}


#-------------- check sudo

if [ "$EUID" -ne 0 ]
  then echo "Permissions denied! Please run as sudo."
  echo -e "\033[0m"
  exit
fi

#-------------- check system
platform=""
arch=""

case "$OSTYPE" in
  darwin*)  platform="darwin" ;; #OSX
  linux*)   platform="linux" ;; #LINUX
esac

case $(uname -m) in
  armv6*)  arch="armv6l" ;;
  armv7*)  arch="armv7l" ;;
  armv8*)  arch="arm64" ;;
  aarch64*)  arch="arm64" ;;
  *)       [[ $(getconf LONG_BIT) = "64" ]] && arch="x64" || arch="x86" ;;
esac

if [[ $platform = "" || $arch = "" ]]
  then echo "This version of $OSTYPE/$(uname -m) is not supported!"
  echo -e "\033[0m"
  exit
else
  run
fi



install.sh

#!/bin/bash


echo -e "\033[0;31m"

#-------------- args
arg1=$1
arg2=$2

#-------------- functions
function config {
   lang="en"
   name_service="ih-v5"

   if [ -n "$arg1" ]; then
     lang="$arg1"
   fi
   if [ -n "$arg2" ]; then
     name_service="$arg2"
   fi

   pwd=$(pwd)
   root="/opt/$name_service"
   temp="$root/temp"
   node="$root/node"
   url="https://github.com/intrahouseio/ih-v5/releases/download/v0.0.0"
}

function clear {
  rm -fr $root
  mkdir -p $root
  mkdir -p $temp
  mkdir -p $node
  cd $root
  
}

function logo {
echo -e "\033[0;34m"
cat <<\EOF
   ██╗██╗  ██╗      ███████╗██╗   ██╗███████╗████████╗███████╗███╗   ███╗███████╗
   ██║██║  ██║      ██╔════╝╚██╗ ██╔╝██╔════╝╚══██╔══╝██╔════╝████╗ ████║██╔════╝
   ██║███████║█████╗███████╗ ╚████╔╝ ███████╗   ██║   █████╗  ██╔████╔██║███████╗
   ██║██╔══██║╚════╝╚════██║  ╚██╔╝  ╚════██║   ██║   ██╔══╝  ██║╚██╔╝██║╚════██║
   ██║██║  ██║      ███████║   ██║   ███████║   ██║   ███████╗██║ ╚═╝ ██║███████║
   ╚═╝╚═╝  ╚═╝      ╚══════╝   ╚═╝   ╚══════╝   ╚═╝   ╚══════╝╚═╝     ╚═╝╚══════╝
                                                                                                                                    
                         Software for Automation Systems                          
-----------------------------------------------------------------------------------
EOF
echo -e "\033[0;31m"
}

function install {
 # echo -e "\033[0;33m"
 # echo -e "INSTALL:\033[0;34m"
  echo -e "\033[0m"
  sudo sudo "$root/node/bin/node" "$root/script.js" $lang $name_service

}

function files {
  echo -e "\033[0;33m"
  echo -e "Get loader:\033[0;34m"

  if [ ! -f "$pwd/script.js" ]; then
      curl -Lo node.tar.gz "$url/node-$platform-$arch.tar.gz"  #script not found
      curl -Lo script.js "https://raw.githubusercontent.com/intrahouseio/ih-v5/main/script.js"
    else
      curl -Lo node.tar.gz "$url/node-$platform-$arch.tar.gz" #script found
      cp "$pwd/script.js" "$root/script.js"
  fi
  
  echo -e "\033[0;33m"
  echo -e "Extract loader:\033[0;34m"
  echo ""

  while true;do echo -n .;sleep 1;done &
  cd ./node
  tar xf ./../node.tar.gz --strip 1 
  cd ./../
  kill $!; trap 'kill $!' SIGTERM
  echo "ok"
}

function run {
  config
  clear
  logo
  files
  # cp "$pwd/script.js" "$root/script.js"
  install
}


#-------------- check sudo

if [ "$EUID" -ne 0 ]
  then echo "Permissions denied! Please run as sudo."
  echo -e "\033[0m"
  exit
fi

#-------------- check system
platform=""
arch=""

case "$OSTYPE" in
  darwin*)  platform="darwin" ;; #OSX
  linux*)   platform="linux" ;; #LINUX
esac

case $(uname -m) in
  armv6*)  arch="armv6l" ;;
  armv7*)  arch="armv7l" ;;
  armv8*)  arch="arm64" ;;
  aarch64*)  arch="arm64" ;;
  *)       [[ $(getconf LONG_BIT) = "64" ]] && arch="x64" || arch="x86" ;;
esac

if [[ $platform = "" || $arch = "" ]]
  then echo "This version of $OSTYPE/$(uname -m) is not supported!"
  echo -e "\033[0m"
  exit
else
  run
fi



install2.sh

#!/bin/bash


echo -e "\033[0;31m"

#-------------- functions
function config {
   pwd=$(pwd)
   name_service="ih-v5"
   root="/opt/$name_service"
   temp="$root/temp"
   node="$root/node"
   url="https://github.com/intrahouseio/ih-v5/releases/download/v0.0.0"
}

function clear {
  rm -fr $root
  mkdir -p $root
  mkdir -p $temp
  mkdir -p $node
  cd $root
  
}

function logo {
echo -e "\033[0;34m"
cat <<\EOF
   ██╗██╗  ██╗      ███████╗██╗   ██╗███████╗████████╗███████╗███╗   ███╗███████╗
   ██║██║  ██║      ██╔════╝╚██╗ ██╔╝██╔════╝╚══██╔══╝██╔════╝████╗ ████║██╔════╝
   ██║███████║█████╗███████╗ ╚████╔╝ ███████╗   ██║   █████╗  ██╔████╔██║███████╗
   ██║██╔══██║╚════╝╚════██║  ╚██╔╝  ╚════██║   ██║   ██╔══╝  ██║╚██╔╝██║╚════██║
   ██║██║  ██║      ███████║   ██║   ███████║   ██║   ███████╗██║ ╚═╝ ██║███████║
   ╚═╝╚═╝  ╚═╝      ╚══════╝   ╚═╝   ╚══════╝   ╚═╝   ╚══════╝╚═╝     ╚═╝╚══════╝
                                                                                                                                    
                         Software for Automation Systems                          
-----------------------------------------------------------------------------------
EOF
echo -e "\033[0;31m"
}

function install {
 # echo -e "\033[0;33m"
 # echo -e "INSTALL:\033[0;34m"
  echo -e "\033[0m"
  sudo sudo "$root/node/bin/node" "$root/script.js"

}

function files {
  echo -e "\033[0;33m"
  echo -e "Get loader:\033[0;34m"

  if [ ! -f "$pwd/script.js" ]; then
      curl -Lo node.tar.gz "$url/node-$platform-$arch.tar.gz"  #script not found
      curl -Lo script.js "https://raw.githubusercontent.com/intrahouseio/ih-v5/main/script2.js"
    else
      curl -Lo node.tar.gz "$url/node-$platform-$arch.tar.gz" #script found
      cp "$pwd/script.js" "$root/script.js"
  fi
  
  echo -e "\033[0;33m"
  echo -e "Extract loader:\033[0;34m"
  echo ""

  while true;do echo -n .;sleep 1;done &
  cd ./node
  tar xf ./../node.tar.gz --strip 1 
  cd ./../
  kill $!; trap 'kill $!' SIGTERM
  echo "ok"
}

function run {
  config
  clear
  logo
  files
  # cp "$pwd/script.js" "$root/script.js"
  install
}


#-------------- check sudo

if [ "$EUID" -ne 0 ]
  then echo "Permissions denied! Please run as sudo."
  echo -e "\033[0m"
  exit
fi

#-------------- check system
platform=""
arch=""

case "$OSTYPE" in
  darwin*)  platform="darwin" ;; #OSX
  linux*)   platform="linux" ;; #LINUX
esac

case $(uname -m) in
  armv6*)  arch="armv6l" ;;
  armv7*)  arch="armv7l" ;;
  armv8*)  arch="arm64" ;;
  aarch64*)  arch="arm64" ;;
  *)       [[ $(getconf LONG_BIT) = "64" ]] && arch="x64" || arch="x86" ;;
esac

if [[ $platform = "" || $arch = "" ]]
  then echo "This version of $OSTYPE/$(uname -m) is not supported!"
  echo -e "\033[0m"
  exit
else
  run
fi



intrascada.sh

#!/bin/bash


echo -e "\033[0;31m"

#-------------- args
arg1=$1
arg2=$2

#-------------- functions
function config {
   lang="en"
   name_service="ih-v5"

   if [ -n "$arg1" ]; then
     lang="$arg1"
   fi
   if [ -n "$arg2" ]; then
     name_service="$arg2"
   fi

   if [ -z $WB_VERSION ]
   then
     root="/opt/$name_service"
   else
     root="/mnt/data/opt/$name_service"
   fi

   pwd=$(pwd)
   assets="$root/assets"
   temp="$root/temp"
   node="$root/node"
   url="https://github.com/intrahouseio/ih-v5/releases/download/v0.0.0"
}

function clear {
  rm -fr $root
  mkdir -p $root
  mkdir -p $assets
  mkdir -p $temp
  mkdir -p $node
  cd $root
  
}

function logo {
echo -e "\033[0;34m"
cat <<\EOF
  ██╗███╗   ██╗████████╗██████╗  █████╗     ███████╗ ██████╗ █████╗ ██████╗  █████╗ 
  ██║████╗  ██║╚══██╔══╝██╔══██╗██╔══██╗    ██╔════╝██╔════╝██╔══██╗██╔══██╗██╔══██╗
  ██║██╔██╗ ██║   ██║   ██████╔╝███████║    ███████╗██║     ███████║██║  ██║███████║
  ██║██║╚██╗██║   ██║   ██╔══██╗██╔══██║    ╚════██║██║     ██╔══██║██║  ██║██╔══██║
  ██║██║ ╚████║   ██║   ██║  ██║██║  ██║    ███████║╚██████╗██║  ██║██████╔╝██║  ██║
  ╚═╝╚═╝  ╚═══╝   ╚═╝   ╚═╝  ╚═╝╚═╝  ╚═╝    ╚══════╝ ╚═════╝╚═╝  ╚═╝╚═════╝ ╚═╝  ╚═╝                                                                               
                                                                                                                                    
                            Software for Automation Systems                          
--------------------------------------------------------------------------------------
EOF
echo -e "\033[0;31m"
}

function install {
  if [ ! -z $WB_VERSION ]
  then
   echo -e "\033[0;33m"
   echo -e "Platform dependencies:\033[0;34m"
   sudo apt-get update && apt-get install -y libatomic1
  fi

  echo -e "\033[0m"
  sudo sudo WB_VERSION=$WB_VERSION "$root/node/bin/node" "$root/script.js" "intrascada" $lang $name_service
}

function files {
  echo -e "\033[0;33m"
  echo -e "Get loader:\033[0;34m"

  if [ ! -f "$pwd/script.js" ]; then
      curl -Lo node.tar.gz "$url/node-$platform-$arch.tar.gz"  #script not found
      curl -Lo script.js "https://raw.githubusercontent.com/intrahouseio/ih-v5/main/script.js"
    else
      curl -Lo node.tar.gz "$url/node-$platform-$arch.tar.gz" #script found
      cp "$pwd/script.js" "$root/script.js"
  fi
  
  echo -e "\033[0;33m"
  echo -e "Extract loader:\033[0;34m"
  echo ""

  while true;do echo -n .;sleep 1;done &
  cd ./node
  tar xf ./../node.tar.gz --strip 1 
  cd ./../
  kill $!; trap 'kill $!' SIGTERM
  echo "ok"
}

function run {
  config
  clear
  logo
  files
  # cp "$pwd/script.js" "$root/script.js"
  install
}


#-------------- check sudo

if [ "$EUID" -ne 0 ]
  then echo "Permissions denied! Please run as sudo."
  echo -e "\033[0m"
  exit
fi

#-------------- check system
platform=""
arch=""

case "$OSTYPE" in
  darwin*)  platform="darwin" ;; #OSX
  linux*)   platform="linux" ;; #LINUX
esac

case $(uname -m) in
  armv6*)  arch="armv6l" ;;
  armv7*)  arch="armv7l" ;;
  armv8*)  arch="arm64" ;;
  aarch64*)  arch="arm64" ;;
  *)       [[ $(getconf LONG_BIT) = "64" ]] && arch="x64" || arch="x86" ;;
esac

if [[ $platform = "" || $arch = "" ]]
  then echo "This version of $OSTYPE/$(uname -m) is not supported!"
  echo -e "\033[0m"
  exit
else
  run
fi



intrahouse.ps1
$Host.UI.RawUI.BackgroundColor = ($bckgrnd = 'Black')
Clear-Host

[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12

#-------------- options
$name_service="ih-v5"

$root_path="$($env:LOCALAPPDATA)\$($name_service)"
$data_path="C:\ProgramData\$($name_service)"

if ( $args ) {
$lang = switch ( $args )
    {
        ru { 'ru' }
        en { 'en' }
        default { 'en' }
    }
} else {
$lang = switch ( $l )
    {
        ru { 'ru' }
        en { 'en' }
        default { 'en' }
    }
}

#-------------- end


#-------------- creation of structures

Remove-Item -Force -Recurse -ErrorAction SilentlyContinue $root_path
New-Item -ItemType Directory -Force -Path $root_path | Out-Null
New-Item -ItemType Directory -Force -Path "$root_path\assets" | Out-Null
New-Item -ItemType Directory -Force -Path "$root_path\tools" | Out-Null
New-Item -ItemType Directory -Force -Path "$root_path\temp" | Out-Null

#-------------- end


#-------------- check root

$currentUser = New-Object Security.Principal.WindowsPrincipal $([Security.Principal.WindowsIdentity]::GetCurrent())
$testadmin = $currentUser.IsInRole([Security.Principal.WindowsBuiltinRole]::Administrator)
if ($testadmin -eq $false) {
[IO.File]::WriteAllLines("$root_path\install.ps1", (New-Object System.Net.WebClient).DownloadString('https://git.io/JOe18'))
$arg="-NoProfile -InputFormat None -ExecutionPolicy Bypass -NoExit -file $root_path\install.ps1 $l"
Start-Process powershell.exe -Verb RunAs -ArgumentList($arg)
exit $LASTEXITCODE
}
#-------------- end


#-------------- check service

if (Get-NetFirewallRule -DisplayName "$name_service" -ErrorAction SilentlyContinue) {
} else {
New-NetFirewallRule -DisplayName "$name_service" -Direction Inbound -Program "$root_path\node-v14.15.1-win-x64\node.exe" -RemoteAddress ANY -Action Allow | Out-Null
}

if (Get-Service -Name "$name_service" -ErrorAction SilentlyContinue) {
cmd /c "SC STOP ihv5.exe" | Out-Null
}
#-------------- end


#-------------- logo

Write-Host -ForegroundColor Blue "





  _____ _   _ _______ _____              _    _  ____  _    _  _____ ______ 
 |_   _| \ | |__   __|  __ \     /\     | |  | |/ __ \| |  | |/ ____|  ____|
   | | |  \| |  | |  | |__) |   /  \    | |__| | |  | | |  | | (___ | |__   
   | | | . ` |  | |  |  _  /   / /\ \   |  __  | |  | | |  | |\___ \|  __|  
  _| |_| |\  |  | |  | | \ \  / ____ \  | |  | | |__| | |__| |____) | |____ 
 |_____|_| \_|  |_|  |_|  \_\/_/    \_\ |_|  |_|\____/ \____/|_____/|______|



                            Software for Automation Systems                          
--------------------------------------------------------------------------------------

"

#-------------- end


#-------------- check dependencies
Write-Host -ForegroundColor DarkYellow "`r`nCheck dependencies:`r`n"
Write-Host "get 7-Zip"
Invoke-WebRequest -Uri "https://github.com/develar/7zip-bin/raw/master/win/ia32/7za.exe" -OutFile "$root_path\tools\7z.exe"

Write-Host "get script"
Invoke-WebRequest -Uri "https://raw.githubusercontent.com/intrahouseio/ih-v5/main/script.js" -OutFile "$root_path\script.js"

#-------------- end


#-------------- download files
Write-Host "get nodeJS"
Invoke-WebRequest -Uri "https://github.com/intrahouseio/ih-v5/releases/download/v0.0.0/node-win-x64.zip" -OutFile "$root_path\node.zip"


#-------------- end


#-------------- deploy
Write-Host -ForegroundColor DarkYellow "`r`nDeploy:`r`n"
cmd /c "$root_path\tools\7z.exe" x -y "$root_path\node.zip" -o"$root_path\"

Remove-Item -Force -Recurse -ErrorAction SilentlyContinue "$root_path\node.zip"
Remove-Item -Force -Recurse -ErrorAction SilentlyContinue "$root_path\install.ps1"

Set-Location "$root_path"
cmd /c "$root_path\node-v14.15.1-win-x64\node.exe" "$root_path\node-v14.15.1-win-x64\node_modules\npm\bin\npm-cli.js" i node-windows --only=prod --no-save --loglevel=error
cmd /c "$root_path\node-v14.15.1-win-x64\node.exe" "$root_path\script.js" intrahouse $lang

#-------------- end


intrascada.ps1
$Host.UI.RawUI.BackgroundColor = ($bckgrnd = 'Black')
Clear-Host

[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12

#-------------- options
$name_service="ih-v5"

$root_path="$($env:LOCALAPPDATA)\$($name_service)"
$data_path="C:\ProgramData\$($name_service)"

if ( $args ) {
$lang = switch ( $args )
    {
        ru { 'ru' }
        en { 'en' }
        default { 'en' }
    }
} else {
$lang = switch ( $l )
    {
        ru { 'ru' }
        en { 'en' }
        default { 'en' }
    }
}

#-------------- end


#-------------- creation of structures

Remove-Item -Force -Recurse -ErrorAction SilentlyContinue $root_path
New-Item -ItemType Directory -Force -Path $root_path | Out-Null
New-Item -ItemType Directory -Force -Path "$root_path\assets" | Out-Null
New-Item -ItemType Directory -Force -Path "$root_path\tools" | Out-Null
New-Item -ItemType Directory -Force -Path "$root_path\temp" | Out-Null

#-------------- end


#-------------- check root

$currentUser = New-Object Security.Principal.WindowsPrincipal $([Security.Principal.WindowsIdentity]::GetCurrent())
$testadmin = $currentUser.IsInRole([Security.Principal.WindowsBuiltinRole]::Administrator)
if ($testadmin -eq $false) {
[IO.File]::WriteAllLines("$root_path\install.ps1", (New-Object System.Net.WebClient).DownloadString('https://git.io/JYpbW'))
$arg="-NoProfile -InputFormat None -ExecutionPolicy Bypass -NoExit -file $root_path\install.ps1 $l"
Start-Process powershell.exe -Verb RunAs -ArgumentList($arg)
exit $LASTEXITCODE
}
#-------------- end


#-------------- check service

if (Get-NetFirewallRule -DisplayName "$name_service" -ErrorAction SilentlyContinue) {
} else {
New-NetFirewallRule -DisplayName "$name_service" -Direction Inbound -Program "$root_path\node-v14.15.1-win-x64\node.exe" -RemoteAddress ANY -Action Allow | Out-Null
}

if (Get-Service -Name "$name_service" -ErrorAction SilentlyContinue) {
cmd /c "SC STOP ihv5.exe" | Out-Null
}
#-------------- end


#-------------- logo

Write-Host -ForegroundColor Blue "





  _____ _   _ _______ _____               _____  _____          _____          
 |_   _| \ | |__   __|  __ \     /\      / ____|/ ____|   /\   |  __ \   /\    
   | | |  \| |  | |  | |__) |   /  \    | (___ | |       /  \  | |  | | /  \   
   | | | . ` |  | |  |  _  /   / /\ \    \___ \| |      / /\ \ | |  | |/ /\ \  
  _| |_| |\  |  | |  | | \ \  / ____ \   ____) | |____ / ____ \| |__| / ____ \ 
 |_____|_| \_|  |_|  |_|  \_\/_/    \_\ |_____/ \_____/_/    \_\_____/_/    \_\



                            Software for Automation Systems                          
--------------------------------------------------------------------------------------

"

#-------------- end


#-------------- check dependencies
Write-Host -ForegroundColor DarkYellow "`r`nCheck dependencies:`r`n"
Write-Host "get 7-Zip"
Invoke-WebRequest -Uri "https://github.com/develar/7zip-bin/raw/master/win/ia32/7za.exe" -OutFile "$root_path\tools\7z.exe"

Write-Host "get script"
Invoke-WebRequest -Uri "https://raw.githubusercontent.com/intrahouseio/ih-v5/main/script.js" -OutFile "$root_path\script.js"

#-------------- end


#-------------- download files
Write-Host "get nodeJS"
Invoke-WebRequest -Uri "https://github.com/intrahouseio/ih-v5/releases/download/v0.0.0/node-win-x64.zip" -OutFile "$root_path\node.zip"


#-------------- end


#-------------- deploy
Write-Host -ForegroundColor DarkYellow "`r`nDeploy:`r`n"
cmd /c "$root_path\tools\7z.exe" x -y "$root_path\node.zip" -o"$root_path\"

Remove-Item -Force -Recurse -ErrorAction SilentlyContinue "$root_path\node.zip"
Remove-Item -Force -Recurse -ErrorAction SilentlyContinue "$root_path\install.ps1"

Set-Location "$root_path"
cmd /c "$root_path\node-v14.15.1-win-x64\node.exe" "$root_path\node-v14.15.1-win-x64\node_modules\npm\bin\npm-cli.js" i node-windows --only=prod --no-save --loglevel=error
cmd /c "$root_path\node-v14.15.1-win-x64\node.exe" "$root_path\script.js" intrascada $lang

#-------------- end


intrahouse.sh

#!/bin/bash


echo -e "\033[0;31m"

#-------------- args
arg1=$1
arg2=$2

#-------------- functions
function config {
   lang="en"
   name_service="ih-v5"

   if [ -n "$arg1" ]; then
     lang="$arg1"
   fi
   if [ -n "$arg2" ]; then
     name_service="$arg2"
   fi

   if [ -z $WB_VERSION ]
   then
     root="/opt/$name_service"
   else
     root="/mnt/data/opt/$name_service"
   fi

   pwd=$(pwd)
   assets="$root/assets"
   temp="$root/temp"
   node="$root/node"
   url="https://github.com/intrahouseio/ih-v5/releases/download/v0.0.0"
}

function clear {
  rm -fr $root
  mkdir -p $root
  mkdir -p $assets
  mkdir -p $temp
  mkdir -p $node
  cd $root
  
}

function logo {
echo -e "\033[0;34m"
cat <<\EOF
  ██╗███╗   ██╗████████╗██████╗  █████╗     ██╗  ██╗ ██████╗ ██╗   ██╗███████╗███████╗
  ██║████╗  ██║╚══██╔══╝██╔══██╗██╔══██╗    ██║  ██║██╔═══██╗██║   ██║██╔════╝██╔════╝
  ██║██╔██╗ ██║   ██║   ██████╔╝███████║    ███████║██║   ██║██║   ██║███████╗█████╗  
  ██║██║╚██╗██║   ██║   ██╔══██╗██╔══██║    ██╔══██║██║   ██║██║   ██║╚════██║██╔══╝  
  ██║██║ ╚████║   ██║   ██║  ██║██║  ██║    ██║  ██║╚██████╔╝╚██████╔╝███████║███████╗
  ╚═╝╚═╝  ╚═══╝   ╚═╝   ╚═╝  ╚═╝╚═╝  ╚═╝    ╚═╝  ╚═╝ ╚═════╝  ╚═════╝ ╚══════╝╚══════╝                                                                               
                                                                                                                                    
                            Software for Automation Systems                          
--------------------------------------------------------------------------------------
EOF
echo -e "\033[0;31m"
}

function install {
  if [ ! -z $WB_VERSION ]
  then
   echo -e "\033[0;33m"
   echo -e "Platform dependencies:\033[0;34m"
   sudo apt-get update && apt-get install -y libatomic1
  fi

  echo -e "\033[0m"
  sudo sudo WB_VERSION=$WB_VERSION "$root/node/bin/node" "$root/script.js" "intrahouse" $lang $name_service
}

function files {
  echo -e "\033[0;33m"
  echo -e "Get loader:\033[0;34m"

  if [ ! -f "$pwd/script.js" ]; then
      curl -Lo node.tar.gz "$url/node-$platform-$arch.tar.gz"  #script not found
      curl -Lo script.js "https://raw.githubusercontent.com/intrahouseio/ih-v5/main/script.js"
    else
      curl -Lo node.tar.gz "$url/node-$platform-$arch.tar.gz" #script found
      cp "$pwd/script.js" "$root/script.js"
  fi
  
  echo -e "\033[0;33m"
  echo -e "Extract loader:\033[0;34m"
  echo ""

  while true;do echo -n .;sleep 1;done &
  cd ./node
  tar xf ./../node.tar.gz --strip 1 
  cd ./../
  kill $!; trap 'kill $!' SIGTERM
  echo "ok"
}

function run {
  config
  clear
  logo
  files
  # cp "$pwd/script.js" "$root/script.js"
  install
}


#-------------- check sudo

if [ "$EUID" -ne 0 ]
  then echo "Permissions denied! Please run as sudo."
  echo -e "\033[0m"
  exit
fi

#-------------- check system
platform=""
arch=""

case "$OSTYPE" in
  darwin*)  platform="darwin" ;; #OSX
  linux*)   platform="linux" ;; #LINUX
esac

case $(uname -m) in
  armv6*)  arch="armv6l" ;;
  armv7*)  arch="armv7l" ;;
  armv8*)  arch="arm64" ;;
  aarch64*)  arch="arm64" ;;
  *)       [[ $(getconf LONG_BIT) = "64" ]] && arch="x64" || arch="x86" ;;
esac

if [[ $platform = "" || $arch = "" ]]
  then echo "This version of $OSTYPE/$(uname -m) is not supported!"
  echo -e "\033[0m"
  exit
else
  run
fi

